### DISCUSSION TITLE
Add do_action before displaying the confirmation page

### DISCUSSION THREAD

**Discussion by Genevieve-K** [impress-org/givewp#7108] [Developer Q&A]
Could you add a `do_action()` when a payment is done and the confirmation page displayed ?
I developed a cart and I want to delete the gifts in the cart when the payment is done. The donation status can be pending for an offline donation or complete. Testing the status is therefore not sufficient. 

I use until now the filter to `'give_get_success_page_uri'` but for Stripe, if there are errors with the payment, the `give_get_success_page_uri()` function is called before creating the payment intent. And this filter is applied even if there is a failure.

I tested to use a filter to `'give_success_page_redirect`', but this work only for legacy gateways. For Stripe, the redirect is in `__invoke()`, in the HandleGatewayPaymentCommand class.

If all gateways use this function, could you add a `do_action()` in this function for example? It's rather urgent.
Thank you by advance.

**Comment by jonwaldstein** [comment]
Hey @Genevieve-K, great suggestions.  I'll need to determine the best place for the do_action to go.  The lifecycle of a donation is a little tricky because gateways are the deciding factor and we rely on them to essentially complete the form how they see fit.  

Sometimes they will tell the framework the payment is complete, other times they will use webhooks to complete the payment, or handle redirecting themselves on the front-end.

To truly know when a donation is complete would be to listen for its status to change through `givewp_donation_updated` or legacy hook `give_update_payment_status`.

To know when a _form_ is complete, I think you are on the right track recommending the confirmation page.

Honestly, I think having an action right before the actual URL request to the confirmation page might be the most reliable way to determine when a form is completed.  In your situation you would probably need to check the confirmation page `$donationId` against a `$donationId` stored in some type of session before clearing anything.  Reason being, the confirmation page is a unique, secure url for each donation and can be visited more than once.

**Comment by Genevieve-K** [comment]
Maybe I can continue to use the `'give_get_success_page_uri'` filter, but in the `create()` function for `CreatePaymentIntent`, an arg named `return_url` is created with the value `give_get_success_page_uri()`, and this arg doesn't seem to be used. The result is that my filter being applied and I delete the gifts in the cart even if an error occurs.

**Reply by Genevieve-K** [reply]
Is the arg named `return_url `useful? Without it, I can continue to use the `'give_get_success_page_uri'` filter.

**Comment by jonwaldstein** [comment]
@Genevieve-K I have temporary solutions for you to try below.  If any of these work we can look into adding some actions to make things easier.

```
/**
 * This could work for v3 forms
 */
add_action('template_redirect', static function() {
    // check for the confirmation page URL
    if (isset($_GET['givewp-route']) && $_GET['givewp-route'] === 'donation-confirmation-receipt-view') {
        // Check if the referer is a valid URL and contains the string 'givewp-route=donation-form-view' which means it came from a GiveWP donation form
        if (filter_var($_SERVER['HTTP_REFERER'], FILTER_VALIDATE_URL) && str_contains('givewp-route=donation-form-view', $_SERVER['HTTP_REFERER']) !== null) {
            // at this point we know the user successfully came from a GiveWP donation form
        }
    }
});
```


```
/**
 * I'm not completely sure but I think this could work for v2 forms (priority would need to be lower than core to work)
 */
add_action('wp_ajax_get_receipt', static function() {
    // might need to get clever here but since this is an ajax request for the receipt - you can probabaly assume the form is complete
    // check LoadTemplate.php handleReceiptAjax() for more inspiration
}, 8);
```

**Reply by Genevieve-K** [reply]
Thank you for your help. Do you see that at this level I still have the $_REQUEST variable?

**Reply by jonwaldstein** [reply]
You would not have access to the POST request if that's what you are referring to.  However, there would be a way to find the donation ID so you could find associated data you need.  What data would you need at this point in the flow?

**Reply by Genevieve-K** [reply]
Yes I can  find the gifts with the donation ID. But it's not simple. I have a data in the post request to get them. I would like to apply my filter before the redirection.

**Reply by jonwaldstein** [reply]
Okay, unfortunately that would not be possible.  Honestly I think you might be better off listening for the donation status to change along with checking what's in the $_POST request. At least then you would know that a donation status is being changed from the original status and checking the available POST request would tell you if it's during the donation form request.

**Reply by Genevieve-K** [reply]
Yes I think it's the best solution. I thought about it.
With the legacy version, I can use the filter` 'give_update_payment_status'` and test the old and new status. 
But with the new version, if I use the hook `'givewp_donation_updated'`, how can I test them? Also add a filter to `'givewp_donation_updating'` to save the old status? Is a donation created with the status 'pending' and then this status is changed?

**Reply by jonwaldstein** [reply]
When the donate button is clicked, GiveWP creates a donation with a `pending` status and hands it over to the gateway.  The gateway will then use the donation details to create a payment using it's api, then update the donation status to either `pending` (again), `processing`, or `complete` ('publish') either directly or through a payment command.  Some gateways will set the donation status to `pending` or `processing` to move forward and will later _complete_ the donation using a webhook.

Here is an example of how you can use the `givewp_donation_updated` action using the model method `isDirty($property)`

```
use Give\Donations\Models\Donation;
use Give\Donations\ValueObjects\DonationStatus;

add_action('givewp_donation_updated', static function (Donation $donation) {
    // using the isDirty method we can check if the status has changed
    if ($donation->isDirty('status')) {
        $allowedDonationStatuses = [
            DonationStatus::PENDING(),
            DonationStatus::PROCESSING(),
            DonationStatus::COMPLETE(),
        ];

        /** @var DonationStatus $originalStatus */
        $originalStatus = $donation->getOriginal('status');
        $newStatus = $donation->status;

        if ($newStatus->isOneOf(...$allowedDonationStatuses)) {
            // we can assume the form is complete by looking at these statuses
            // gateways will often update the status to pending, processing or complete before the form is complete
            // some gateways will return later with a webhook to update the status to complete
            
            // do something
            give_update_meta($donation->id, 'my_custom_meta_key', 'my_custom_meta_value');
        }
    }
});
```

**Reply by Genevieve-K** [reply]
Waouh! isDirty() is a magic function!
I added an action to 'give_success_page_redirect' for legacy gateways. I tested with several gateways (offline, Stripe...). It seems good.
Thank you!

**Reply by jonwaldstein** [reply]
@Genevieve-K glad to hear!

**Reply by Genevieve-K** [reply]
Thank you a lot for your help. I have now 3 filters: for 'give_success_page_redirect', 'givewp_donation_updated' and 'give_update_payment_status'. 
I still have a case where the gifts remain in the cart. It's for a recurring donation with PayPal Commerce. In the donation confirmation, the status is 'pending', and a few minutes later, after PayPal responses, the status changed to 'Complete'. But it's too late., the donation confirmation is still displayed.
Which filter do you advice? Perhaps when the PayPal status changes to  'APPROVAL_PENDING' ? an action to 'give_recurring_post_create_payment_profiles' ?

**Reply by jonwaldstein** [reply]
@Genevieve-K

It looks like PayPal Commerce sets the donation status `processing` and then will complete the status via webhook.

I think in some cases for recurring, you could check the subscription status property instead of the donation status property - basically the same way as above but using `givewp_subscription_updated`.

```
add_action('givewp_subscription_updated', static function (Give\Subscriptions\Models\Subscription $subscription) {
    if ($subscription->isDirty('status')) {
        $allowedSubscriptionStatuses = [
            Give\Subscriptions\ValueObjects\SubscriptionStatus::PENDING(),
            Give\Subscriptions\ValueObjects\SubscriptionStatus::COMPLETED(),
            Give\Subscriptions\ValueObjects\SubscriptionStatus::ACTIVE(),
        ];

        /** @var Give\Subscriptions\ValueObjects\SubscriptionStatus $originalStatus */
        $originalStatus = $subscription->getOriginal('status');
        $newStatus = $subscription->status;

        if ($newStatus->isOneOf(...$allowedSubscriptionStatuses)) {
            // do something
        }

    }
});
```

**Reply by Genevieve-K** [reply]
Thank you. I try with these status:
```
		$allowedSubscriptionStatuses = [
			SubscriptionStatus::PENDING(),
			SubscriptionStatus::COMPLETED(),
			SubscriptionStatus::ACTIVE(),
		];
```
It's difficult to test because PayPal donations work from time to time.

**Comment by Genevieve-K** [comment]
I was finally able to test it. My gift is still in the cart. I don't think that subscription status changes until PayPal responds via the webhook.
In `PaymentGateways/PayPalCommerce/Repositories/Subscription.php`, there are some functions as `updateStatus()` for a `$paypalSubscriptionId`, `isSuccessful()`. But I don't se how to exploit them.

**Reply by Genevieve-K** [reply]
@jonwaldstein I see in class `SubscriptionProcessingHandler `that the donation status is set to `DonationStatus::PROCESSING()`. But this class seems to be invoked only for legacy gateways. Perhaps the same thing when processing a PayPal recurring donation is a good idea. I think that it's not possible for now to use a filter.
For the moment, it works for test and Stripe recurring donations. I have to test with RazorPay.
